Store: Add a "product search" component

ryelle opened the pull request:
This PR creates a new component, `ProductSearch`, which displays a `SearchCard` and is used to search through a site's projects. This will eventually be used on the order edit & creation screens, and might be useful elsewhere.

![screen shot 2017-09-27 at 6 54 06 pm](https://user-images.githubusercontent.com/541093/30941784-4397d2a6-a3b5-11e7-9be4-0d0ad97f8f7e.png)

In addition to returning matching products, it auto-expands variations into selectable items:

![screen shot 2017-09-27 at 6 48 49 pm](https://user-images.githubusercontent.com/541093/30941656-91017e9e-a3b4-11e7-8a19-7069314aaf0e.png)

It's up to the parent component to pass a prop `onSelect`, which fires when a result is clicked.

Currently the only required prop is the `onSelect` function, but I could also see maybe wanting to control whether variations are included - let me know if that would be useful 🙂 

To test:

- This is currently used on the create order page, so you can visit there to test it: `/store/order/:site`
- Try searching a few different products, things that don't exist, products with and without variations, etc.
- Clicking a result selects it, but nothing is being done with the selected object yet, so it should just close the result set.
- Keyboard navigation (tabbing to items, using space or enter to select) should also work like clicking

timmyc commented on client/extensions/woocommerce/components/product-search/item.js:
```diff
@@ -0,0 +1,90 @@
+import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
+import { getVariationsForProduct } from 'woocommerce/state/sites/product-variations/selectors';
+
+class ProductItem extends Component {
```
Missing `propTypes` declaration here.

timmyc commented on client/extensions/woocommerce/components/product-search/item.js:
```diff
@@ -0,0 +1,90 @@
+		return {
+			productId,
+			siteId: site && site.ID,
+			variations: getVariationsForProduct( state, productId ) || [],
```
Looks like `getVariationsForProduct` will return null if nothing is found - so we could consider omitting defaulting to an empty array here, and in line 66 just checking for truthiness of `variations` should work. Either way works fine, just a thought.

timmyc commented on client/extensions/woocommerce/components/product-search/results.js:
```diff
@@ -0,0 +1,55 @@
+
+class ProductSearchResults extends Component {
+	static propTypes = {
+		search: PropTypes.string.isRequired,
```
missing a few props here: `isLoaded, isLoading, onSelect, products, translate` - I know we talked about props at the GM but couldn't remember if we wanted to be super explicit with them or not - I'm a fan though 😆

ryelle replied:
I'll grab the rest of your feedback tomorrow, but for this one -- I generally only define the passed-in PropTypes (which does mean I'm still missing `onSelect`), but things like `translate` or connected props seem less useful.

spen replied:
If you don't mind me adding my 2c, I'd say that strictly all cases where a component relies on a value or function _should_ be covered by either `propTypes` or `defaultProps`. If `translate` wasn't passed in we'd see runtime errors. That said, we aren't exporting this component without enhancements (`localize` & `connect`) so until we do (for testing perhaps) we could be a little less strict :)

ryelle commented:
Good catch on the scrollbars, I always forget to test that setting 🙂  Maybe @kellychoffman or @jameskoster can help out with the CSS there?

Agreed about the search, it's not totally intuitive. It would also be cool if it searched into variations too, for example I have a t-shirt with multiple colors, but if search "green" it wouldn't show up. That's an enhancement for the API side though.

The focus state of the first element is always going to be hidden due to how the container class works, which is also why you see the small gap between the scroll bars - I've added padding into the container so the focus state isn't hidden.

And lastly, another good catch on using the variation image 😄 I'll check that out tomorrow as well.

spen commented on client/extensions/woocommerce/components/product-search/item.js:
```diff
@@ -0,0 +1,90 @@
+	};
+
+	renderItem = product => {
+		const featuredImage = product.images && product.images[ 0 ];
```
What do you think to using `_.get` here?
I put together a [quick demo](https://runkit.com/spen/59cc49e6a000c80012b0ac25) of how the case would be handled:

spen commented on client/extensions/woocommerce/components/product-search/item.js:
```diff
@@ -0,0 +1,90 @@
+		}
+	}
+
+	onClick = () => {
```
We [should be avoiding](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#react-components) using names like `onClick` for handlers, perhaps `handleClick` may be better suited?

spen commented on client/extensions/woocommerce/components/product-search/item.js:
```diff
@@ -0,0 +1,90 @@
+
+		return {
+			productId,
+			siteId: site && site.ID,
```
I would encourage `get` again here, though it's a very minor change so FFTI :)

ryelle commented on client/extensions/woocommerce/app/order/order-create/index.js:
```diff
@@ -18,9 +18,12 @@ import { getSelectedSiteWithFallback } from 'woocommerce/state/sites/selectors';
 import { ProtectFormGuard } from 'lib/protect-form';
 import SectionHeader from 'components/section-header';
 
+const noop = () => {};
```
If this were going to stay here I would, but we're about to redo this entire section (it's feature-flagged anyway), I just needed someplace to put the component for testing :)

spen replied:
Makes sense :) TY for clarifying!

timmyc replied:
I thought about mentioning that too, but figured it was temporary 😆

timmyc commented:
@jameskoster really like the loading state design above 🌟

ryelle commented:
I think I've gotten everyone's feedback addressed now, and I've rebased against master. The new loading state and "nothing found" states are also included now. I've added variation images when they exist, and tweaked the scroll CSS a little so only the vertical scrollbar should appear.

allendav commented:
> Keyboard navigation (tabbing to items, using space or enter to select) should also work like clicking

Tabbing kinda works, but it takes two TAB presses to see the first item actually display an outline (and the outline isn't quite right on the first item)

![tabbing](https://user-images.githubusercontent.com/1595739/30984231-9dcf82c8-a441-11e7-8346-9ea291bcbcb3.gif)

allendav commented on client/extensions/woocommerce/components/product-search/index.js:
```diff
@@ -0,0 +1,81 @@
+	handleSearch = query => {
+		const { siteId } = this.props;
+
+		if ( trim( query ) === '' ) {
```
Optional. Yoda.

allendav commented on client/extensions/woocommerce/components/product-search/index.js:
```diff
@@ -0,0 +1,81 @@
+	};
+
+	handleSearch = query => {
+		const { siteId } = this.props;
```
Optional. Consider debouncing this. As the number of products on a site grows, I suspect these search queries could take some time/resources to complete

allendav replied:
Nevermind. TIL about delaySearch :P

allendav commented on client/extensions/woocommerce/components/product-search/results.js:
```diff
@@ -0,0 +1,103 @@
+					<Gridicon icon="info-outline" />
+				</div>
+				<div className="product-search__label">
+					<div className="product-search__name">{ translate( 'No products found' ) }</div>
```
Nit: If this happens, probably don't want cursor pointer?

![img_0258](https://user-images.githubusercontent.com/1595739/30984860-77e4b2c0-a443-11e7-8c87-5e820cf81025.JPG)

kellychoffman commented:
> Clicking a result selects it, but nothing is being done with the selected object yet, so it should just close the result set.

Is hover and active states out of scope for this PR?

ryelle commented:
> Are hover and active states out of scope for this PR?

No, do you want it to do something more than the cursor-mouse on hover?

kellychoffman commented:
Sorry, failed to get to the styling of this yesterday. Was thinking we could adopt the same hover/active states as the SelectDropdown, since its most similar. https://wpcalypso.wordpress.com/devdocs/design/select-dropdown

jameskoster commented:
Added hover styles to match the theme search results;

![products](https://cldup.com/6uFpPfn9ac.png)

Themes;

![themes](https://cldup.com/FLB56fTMce.png)

ryelle commented:
@jameskoster Thanks for the style updates. I've just added the code to highlight the search string inside the product name, so it looks like this now:

<img width="698" alt="screen shot 2017-10-02 at 11 20 40 am" src="https://user-images.githubusercontent.com/541093/31084983-eb12fd60-a763-11e7-9d73-37cfd7b19555.png">

ryelle commented:
I've rebased, and will merge this in a few hours, I think we've addressed all the feedback - if I missed anything, let me know 🙂
